File: App/Private Messages/NewMessageChecker.m
//  NewMessageChecker.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "NewMessageChecker.h"
#import "AwfulForumsClient.h"
@import UIKit;

#import "Awful-Swift.h"

@interface NewMessageChecker ()

@property (assign, nonatomic) NSInteger unreadCount;
@property (strong, nonatomic) NSTimer *timer;

@end

@implementation NewMessageChecker

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (id)init
{
    if ((self = [super init])) {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(applicationWillEnterForeground:)
                                                     name:UIApplicationWillEnterForegroundNotification
                                                   object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(applicationDidEnterBackground:)
                                                     name:UIApplicationDidEnterBackgroundNotification
                                                   object:nil];
        [self startTimer];
    }
    return self;
}

- (NSInteger)unreadCount
{
    return [[NSUserDefaults standardUserDefaults] integerForKey:UnreadMessageCountKey];
}

- (void)setUnreadCount:(NSInteger)unreadCount
{
    [[NSUserDefaults standardUserDefaults] setInteger:unreadCount forKey:UnreadMessageCountKey];
    [[NSNotificationCenter defaultCenter] postNotificationName:NewMessageCheckerUnreadCountDidChangeNotification
                                                        object:self
                                                      userInfo:@{NewMessageCheckerUnreadCountKey: @(unreadCount)}];
}

- (void)decrementUnreadCount
{
    if (self.unreadCount > 0) {
        self.unreadCount--;
    }
}

static NSString * const UnreadMessageCountKey = @"AwfulUnreadMessages";

+ (instancetype)sharedChecker
{
    static NewMessageChecker *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [self new];
    });
    return instance;
}

- (void)applicationWillEnterForeground:(NSNotification *)note
{
    [self refreshIfNecessary];
    [self startTimer];
}

- (void)applicationDidEnterBackground:(NSNotification *)note
{
    [_timer invalidate];
    _timer = nil;
}

- (void)startTimer
{
    NSTimeInterval interval = [[[RefreshMinder sharedMinder] suggestedDateToRefreshNewPrivateMessages] timeIntervalSinceNow];
    _timer = [NSTimer scheduledTimerWithTimeInterval:interval target:self selector:@selector(timerDidFire:) userInfo:nil repeats:NO];
}

- (void)timerDidFire:(NSTimer *)timer
{
    _timer = nil;
    [self refreshIfNecessary];
}

- (void)refreshIfNecessary
{
    if ([RefreshMinder sharedMinder].shouldRefreshNewPrivateMessages) {
        __weak __typeof__(self) weakSelf = self;
        [[AwfulForumsClient client] countUnreadPrivateMessagesInInboxAndThen:^(NSError *error, NSInteger unreadCount) {
            __typeof__(self) self = weakSelf;
            if (error) {
                NSLog(@"%s error checking for new private messages: %@", __PRETTY_FUNCTION__, error);
                return;
            }
            self.unreadCount = unreadCount;
            [[RefreshMinder sharedMinder] didRefreshNewPrivateMessages];
        }];
    }
}

@end

NSString * const NewMessageCheckerUnreadCountDidChangeNotification = @"Awful.NewMessageCheckerUnreadCountDidChangeNotification";
NSString * const NewMessageCheckerUnreadCountKey = @"unreadCount";

File: App/Private Messages/NewMessageChecker.swift
//  NewMessageChecker.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore
import Foundation

/// Periodically checks for new private messages in the logged-in user's inbox and posts notifications.
final class NewMessageChecker: NSObject {
    static let sharedChecker = NewMessageChecker()
    fileprivate var timer: Timer?
    
    override init() {
        super.init()
        
        NotificationCenter.default.addObserver(self, selector: #selector(applicationWillEnterForeground), name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationDidEnterBackground), name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
    }
    
    var unreadCount: Int {
        get { return UserDefaults.standard.integer(forKey: unreadCountKey) }
        set {
            UserDefaults.standard.set(newValue, forKey: unreadCountKey)
            let userInfo = [NewMessageChecker.notificationUnreadCountKey: newValue]
            NotificationCenter.default.post(name: Notification.Name(rawValue: NewMessageChecker.didChangeNotification), object: self, userInfo: userInfo)
        }
    }
    
    func decrementUnreadCount() {
        guard unreadCount > 0 else { return }
        unreadCount -= 1
    }
    
    fileprivate func startTimer() {
        timer?.invalidate()
        let interval = RefreshMinder.sharedMinder.suggestedRefreshDate(.newPrivateMessages).timeIntervalSinceNow
        timer = Timer.scheduledTimerWithInterval(interval, handler: { [weak self] timer in
            self?.refreshIfNecessary()
        })
    }
    
    func refreshIfNecessary() {
        guard RefreshMinder.sharedMinder.shouldRefresh(.newPrivateMessages) else { return }
        _ = ForumsClient.shared.countUnreadPrivateMessagesInInbox { [weak self] (error: Error?, unreadCount) in
            if let error = error {
                print("\(#function) error checking for new private messages: \(error)")
                return
            }
            if let unreadCount = unreadCount {
                self?.unreadCount = unreadCount
            }
            RefreshMinder.sharedMinder.didRefresh(.newPrivateMessages)
        }
    }
    
    static let didChangeNotification = "Awful.NewMessageCheckerUnreadCountDidChangeNotification"
    static let notificationUnreadCountKey = "unreadCount"
    
    @objc fileprivate func applicationWillEnterForeground(_ notification: Notification) {
        refreshIfNecessary()
        startTimer()
    }
    
    @objc fileprivate func applicationDidEnterBackground(_ notification: Notification) {
        timer?.invalidate()
        timer = nil
    }
}

private let unreadCountKey = "AwfulUnreadMessages"
